src/Smalot/PdfParser/Font.php

```diff
@@ -381,86 +393,182 @@ public function decodeText(array $commands): string
 
-                $text = $result;
+                $result .= $char;
             }
-        } elseif ($this->has('Encoding') && $this->get('Encoding') instanceof Encoding) {
-            /** @var Encoding $encoding */
-            $encoding = $this->get('Encoding');
-            $unicode = mb_check_encoding($text, 'UTF-8');
-            $result = '';
-            if ($unicode) {
-                $chars = preg_split(
-                        '//s'.($unicode ? 'u' : ''),
-                        $text,
-                        -1,
-                        \PREG_SPLIT_DELIM_CAPTURE | \PREG_SPLIT_NO_EMPTY
-                );
-
-                foreach ($chars as $char) {
-                    $dec_av = hexdec(bin2hex($char));
-                    $dec_ap = $encoding->translateChar($dec_av);
-                    $result .= self::uchr($dec_ap ?? $dec_av);
-                }
-            } else {
-                $length = \strlen($text);
 
-                for ($i = 0; $i < $length; ++$i) {
-                    $dec_av = hexdec(bin2hex($text[$i]));
-                    $dec_ap = $encoding->translateChar($dec_av);
-                    $result .= self::uchr($dec_ap ?? $dec_av);
-                }
-            }
             $text = $result;
-        } elseif ($this->get('Encoding') instanceof Element &&
-                  $this->get('Encoding')->equals('MacRomanEncoding')) {
-            // mb_convert_encoding does not support MacRoman/macintosh,
-            // so we use iconv() here
-            $text = iconv('macintosh', 'UTF-8', $text);
-        } elseif (!mb_check_encoding($text, 'UTF-8')) {
-            // don't double-encode strings already in UTF-8
-            $text = mb_convert_encoding($text, 'UTF-8', 'Windows-1252');
         }
 
         return $text;
     }
+
+    /**
+     * Decode content by any type of Encoding (dictionary's item) instance.
+     */
+    private function decodeContentByEncoding(string $text): ?string
+    {
+        $encoding = $this->get('Encoding');
+
+        // When Encoding referenced by object id but object itself not contains `/Type /Encoding` in it's dictionary.
+        if ($encoding instanceof PDFObject) {
+            $encoding = $this->getInitializedEncodingByPdfObject($encoding);
+        }
+
+        // When Encoding contains `/Type /Encoding` in it's dictionary.
+        if ($encoding instanceof Encoding) {
+            return $this->decodeContentByEncodingEncoding($text, $encoding);
+        }
+
+        // When Encoding is just String
+        if ($encoding instanceof Element) { //todo: ElementString class must by used?
+            return $this->decodeContentByEncodingElement($text, $encoding);
+        }
+
+        // Encoding has unintended type.
+        $encodingClassName = \get_class($encoding);
+        throw new LogicException("Unknown encoding instance type: {$encodingClassName}");
+    }
+
+    private function getInitializedEncodingByPdfObject(PDFObject $PDFObject): Encoding
```

Comment:
Why did you add this function? It is only used once.

Reply:
> Why did you add this function?

Reason for this function is described in method where it used.

https://github.com/likemusic/pdfparser/blob/21c4797009c23979005fc6fb1ee358c2abd8dfb8/src/Smalot/PdfParser/Font.php#L465-L468

Should I add some comments for this method? What exactly should I add to the comment?

Reply:
> It is only used once.

Is there anything wrong with that?

Reply:
> Should I add some comments for this method? What exactly should I add to the comment?

Its usually the way around: each function should be described in its function header. If one uses a function, then there could be more information about the why's, what's etc. Having a descriptive function header also benefits users, who use an IDE. They will get information when using a function beforehand.

> Is there anything wrong with that?

There is a good statement about usage of functions [here](https://www.cs.utah.edu/~zachary/computing/lessons/uces-10/uces-10/node11.html#:~:text=The%20first%20reason%20is%20reusability,program%2C%20which%20saves%20you%20work.&text=Another%20aspect%20of%20reusability%20is,different%20(and%20separate)%20programs.). It talks about reusability and abstraction. In the respective [code line](https://github.com/likemusic/pdfparser/blob/21c4797009c23979005fc6fb1ee358c2abd8dfb8/src/Smalot/PdfParser/Font.php#L467) you introduce a function which uses another introduced function which just initialize an instance. Combined with the fact that you only using them once, that is over engineered in my opinion. It suggests complexity/structures which are not needed.

Please combined them.

Reply:
> each function should be described in its function header

Imho, the function should be described if the description has any sense, any benefit.

I don't see any sense to add comments which just repeat the name of the function, like below:

```
/**
* Get user by user id
* 
* @param int $userId
* @return User
**/

public function getUserById(int $userId): User
```

Reply:
> There is a good statement about usage of functions here. It talks about reusability and abstraction.

It's doesn't mean that you should use function only if you use both reasons. Just one of them is enough to use a function.

Reply:
> you introduce a function which uses another introduced function which just initialize an instance. Combined with the fact that you only using them once, that is over engineered in my opinion.

In this case, it's just abstraction. 
In my code, I adhere to the following rule:
- If created by function object must be configured, then it should be split to function which just creates an object.

It helps to understand just by function name if something else happens during creating an object. Also if you'll need to have an object with another configuration, you don't need to refactor it in the future.

Reply:
> It suggests complexity/structures which are not there.

As I argued earlier, splitting function to other do not grow complexity. It just adds some private methods to a class, which do not affect understanding the algorithm of the public methods in any way.

Otherwise, when you use inline many lines of code instead of one line function call, it very much increases function size and cognitive load to understand what a function really do.

Reply:
> Imho, the function should be described if the description has any sense, any benefit.
> I don't see any sense to add comments which just repeat the name of the function, like below:

You are right, I wouldn't bother to add function headers to getter-functions either. What I meant and referring to were functions which you introduced and do more than just returning a value. When writing code which is used by others, I ask myself "What do others need to know about my functions to properly use them?".
---

I think we got a good idea what the other person thinks about the topic. Also, I don't wanna start a philosophical discussion here. I follow your arguments here and therefore, this conversation is resolved. Nevertheless, don't hesitate to add further comments.